ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveringCustodyImpl.java

```diff
@@ -299,28 +300,30 @@ public SafeFuture<Optional<DataColumnSidecar>> getCustodyDataColumnSidecarByRoot
   @Override
   public SafeFuture<Void> onNewValidatedDataColumnSidecar(
       final DataColumnSidecar dataColumnSidecar) {
-    createOrUpdateRecoveryTaskForDataColumnSidecar(
-        DataColumnSlotAndIdentifier.fromDataColumn(dataColumnSidecar));
+    createOrUpdateRecoveryTaskForDataColumnSidecar(dataColumnSidecar);
     return delegate.onNewValidatedDataColumnSidecar(dataColumnSidecar);
   }
 
-  private synchronized void createOrUpdateRecoveryTaskForDataColumnSidecar(
-      final DataColumnSlotAndIdentifier identifier) {
-    if (recoveryTasks.containsKey(identifier.getSlotAndBlockRoot())) {
-      final RecoveryTask existing = recoveryTasks.get(identifier.getSlotAndBlockRoot());
-      existing.existingColumnIds().add(identifier);
-      maybeStartRecovery(existing);
-    } else {
-      final ConcurrentHashSet<DataColumnSlotAndIdentifier> identifiers = new ConcurrentHashSet<>();
-      identifiers.add(identifier);
+  private void createOrUpdateRecoveryTaskForDataColumnSidecar(final DataColumnSidecar sidecar) {
+    if (!recoveryTasks.containsKey(sidecar.getSlotAndBlockRoot())) {
+      final Map<DataColumnSlotAndIdentifier, DataColumnSidecar> sidecars =
+          new ConcurrentHashMap<>();
+      sidecars.put(DataColumnSlotAndIdentifier.fromDataColumn(sidecar), sidecar);
       final RecoveryTask recoveryTask =
           new RecoveryTask(
               new AtomicReference<>(null),
```

Comment:
can we give it a better value than a null reference?

Reply:
It at least appears we store the whole block ,and we only care about the `slotAndBlockRoot`.
given we can get SlotAndBlockRoot from the sidecar, i'd probably lean towards just having a `SlotAndBlockRoot` attribute of the recoveryTask, and not store the block at all. then we'd no longer need the logic to set it either, and it can never be null.

Reply:
Nice catch. As far as we cannot start recovery without 64 sidecars, I've removed onBlock handler at all and changed this field to `slotAndBlockRoot` without concurrency protection as we set it only once and that's all